Modules/posixmodule.c

```diff
@@ -3747,55 +3740,52 @@ os__getfinalpathname_impl(PyObject *module, path_t *path)
         /* FILE_FLAG_BACKUP_SEMANTICS is required to open a directory */
         FILE_FLAG_BACKUP_SEMANTICS,
         NULL);
+    Py_END_ALLOW_THREADS
 
     if (hFile == INVALID_HANDLE_VALUE) {
-        err = "CreateFileW";
-        goto done1;
+        return win32_error_object("CreateFileW", path->object);
```

Comment:
I might be misremembering this, but I'm fairly sure we need to capture `GetLastError` within `Py_END_ALLOW_THREADS` to avoid losing the information if the lock acquisition overwrites it.

Reply:
I questioned that a while ago and searched `Py_END_ALLOW_THREADS` in `Modules/posixmodule.c`. Every piece of code I looked at didn't bother to preserve the Windows last error across `Py_END_ALLOW_THREADS`, so I assumed it's OK.

Now that I've actually checked, `Py_END_ALLOW_THREADS` does make an effort to preserve POSIX `errno`, but it does nothing special for Windows. And the Windows last error may change in that macro, for example, because `take_gil` calls `SleepConditionVariableSRW` which may timeout and is documented to set the last error to `ERROR_TIMEOUT` in this case.  The only code that cares about the last error is TLS-related (in `Python/thread_nt.h`), but despite the comments there (e.g. `PyThread_tss_get`) suggest that the code preserves the last error specifically for `Py_END_ALLOW_THREADS`, it's clearly not enough.

Should I change this PR to avoid losing the last error in `nt._getfinalpathname` only, or is it better to address the issue globally in a separate PR by changing `Py_END_ALLOW_THREADS` macro internals?

Reply:
Correction: I've rechecked, and it turns out that Python unconditionally uses older XP+ synchronization primitives instead of Vista+ primitives (introduced in e75ff35af2b6), so `SleepConditionVariableSRW` is not called. The set of WinAPIs called from `Py_END_ALLOW_THREADS` and not wrapped with last-error-preservation code consists of `EnterCriticalSection`, `LeaveCriticalSection`, `WaitForSingleObjectEx` and `ReleaseSemaphore`. None of them are documented to preserve the last error on success, but at least  `WaitForSingleObjectEx` reports timeout via its return value instead of the last error. So, while there *might* be a chance that `Py_END_ALLOW_THREADS` doesn't clobber the last error in practice, I'd still not rely on this and fix the macro.

Reply:
Great, thanks for the thoroughness! Nothing blocking this patch, so I'll merge it and get the backports started and file a new issue (or resurrect an old one) about preserving the error.

(FWIW, _all_ Win32 APIs should preserve the error value on success, or they have a bug, and all are allowed and should overwrite the last error value on failure. And we currently aren't using the new primitives because of issues whenever we enable them - there are issues on bpo about this if you wanted to look them up.)